src/protocol/sort/secureapplyinv.rs

```diff
@@ -0,0 +1,172 @@
+use crate::{
+    error::BoxError,
+    field::Field,
+    helpers::mesh::{Gateway, Mesh},
+    protocol::{
+        context::ProtocolContext,
+        sort::ApplyInvStep::{ShuffleInputs, ShufflePermutation},
+        IPAProtocolStep::{self, Sort},
+        RecordId,
+        SortStep::ApplyInv,
+    },
+    secret_sharing::Replicated,
+};
+use embed_doc_image::embed_doc_image;
+use futures::future::try_join_all;
+use permutation::Permutation;
+
+use super::{
+    apply::apply_inv, compose_ipa_step, concat_two_ipa_steps, shuffle::Shuffle, ApplyInvStep,
+    SortStep,
+};
+use tokio::try_join;
+
+/// This is APPLYINV(Algorithm 4) described in <https://eprint.iacr.org/2019/695.pdf>.
+/// This is a protocol that applies the inverse of a shared permutation to a shared-vector representation.
+/// Input: Each helpers know their own secret shares of input and permutation
+/// Output: At the end of the protocol, all helpers receive inputs after the permutation is applied
+#[derive(Debug)]
+pub struct SecureApplyInv<'a, F> {
+    input: &'a mut Vec<Replicated<F>>,
+    permutation: &'a mut Vec<Replicated<F>>,
+}
+
+impl<'a, F: Field> SecureApplyInv<'a, F> {
+    #[allow(dead_code)]
+    pub fn new(input: &'a mut Vec<Replicated<F>>, permutation: &'a mut Vec<Replicated<F>>) -> Self {
+        Self { input, permutation }
+    }
+    #[embed_doc_image("secureapplyinv", "images/sort/secureapplyinv.png")]
+    /// In a nutshell, this algorithm applies a permutation on the inputs. The permutation is not available in clear and
+    /// we do not want to reveal the actual permutation to any of the helpers while applying permutation.
+    /// Steps
+    /// ![Secure Apply Inv steps][secureapplyinv]
+    /// 1. Secret shared permutation is shuffled with a permutation
+    /// 2. Secret shared value is shuffled using the same permutation as 1
+    /// 3. The permutation is revealed
+    /// 4. All helpers call `apply_inv` to apply the permutation locally.
+    #[allow(dead_code)]
+    pub async fn execute<M: Mesh, G: Gateway<M, IPAProtocolStep>>(
+        &mut self,
+        ctx: &ProtocolContext<'_, G, IPAProtocolStep>,
+    ) -> Result<(), BoxError> {
+        // 1. Shuffle permutations
+        let shuffle_perm_fn = compose_ipa_step!(ShufflePermutation, ApplyInv, Sort);
+        let mut shuffle_permutation = Shuffle::new(self.permutation, shuffle_perm_fn);
+        let h0_future = shuffle_permutation.execute(ctx);
+
+        // 2. Shuffle inputs
+        let shuffle_input_fn = compose_ipa_step!(ShuffleInputs, ApplyInv, Sort);
+        let mut shuffle_input = Shuffle::new(self.input, shuffle_input_fn);
+        let h1_future = shuffle_input.execute(ctx);
```

Comment:
ApplyInv needs to call shuffle two times once for shuffling inputs and another for shuffling permutation. They both need to use the same sequence combination to calculate the shuffle. However, the infra needs a unique step for each communication requested. With each unique step comes a new prss space which will generate a new sequence not shared between input and permutation.
We would need the exact functionality for COMPOSE as well which will share the sequence between shuffle and unshuffle.
What is the best way to do this? @martinthomson @akoshelev

Reply:
```rust
enum ApplyInvStep {
  ShuffleInput(ShuffleStep),
  ShufflePermutation(ShuffleStep),
  /* plus any other */
}
```

Then you need to work through the integration of the pieces into a single numbering space:

```rust
impl SpaceIndex for ApplyInvStep { 
  const MAX: usize = ShuffleStep::MAX * 2;
  fn as_usize(&self) -> usize {
    match self {
      ShuffleInput(s) => s.as_usize(),
      ShufflePermutation(s) => ShuffleStep::MAX + s.as_usize(),
    }
  }
}
```

Reply:
I am not 100% sure i understand Space completely. How would changing ShuffleInput(s) => 0 to s.usize() help here? 
Also, do the types in a Step (e.g. ShuffleInput and ShufflePermutation in ApplyInvStep) share same prss random seed? i.e. will generate_value(INDEX) called for ShuffleInput give same output as generate_value(INDEX) called for ShufflePermutation? @martinthomson

Reply:
The goal is to ensure that every step that is executed during a query gets a different seed.  For that, you need a unique value for each, even if it is doing the same work as another step.  For instance, we multiply many times, but each multiplication needs a different identifier.

What this does is - at a local level - give each discrete step a different number.  It also lets you compose simple processes (say you just want to shuffle on its own) into more complex processes (applyinv with two shuffles).

Reply:
Is it intentional that each step in the IPAProtocol should share the same prss? For now, all of the steps seem to be having the same prss (initialized by same random number).
For shuffle, we would need to have different random numbers across 3 steps. One way to handle that can be to call generate_values with different index for each step.

Reply:
The idea here is to have a single Participant (naming likely needs work).  From that, multiple PrssSpace instances are created, one for each Step that is executed.  The way that we ensure that each PrssSpace is different is by providing a unique identifier and that is the responsibility of the SpaceIndex trait that IPAProtocolStep implements.

In a full execution of the protocol, there will be multiple places where we need a PRSS instance (PrssSpace).  There will be multiple shuffles and multiple multiplications.  Each of these will need a unique identifier so that the PrssSpace is different.

Ultimately, we need to provide unique random values for every single invocation of the PRSS (or PrssSpace).  For that, we need a unique identifier for the step and a unique identifier for the record that is being processed in that step.

(Note that in testing, you can use a smaller or more narrowly scoped SpaceIndex implementation that provides unique identifiers for each step.  Our current implementation here is lacking as it doesn't provide unique identifiers for each.)